### PR TITLE
image viewer

### PR DESCRIPTION
**Pull request by kson1128**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,16 @@
-# osd_zoom_level
+# Image Viewer
+
+## Overview
+
+This is an interactive image viewer built with the OpenSeadragon library. The viewer loads an image from a specified URL and enables users to zoom in and out at predefined magnification levels.
```
Comment (chunloy) [inline README.md:5]: Could maybe mention that this is an enhanced viewer with a preset widget for quick zooming?

Comment (chunloy) [reply README.md:5]: Also add a section with a demo, screenshots or gif should be fine

File: express.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  })
+);
+
+app.get('/', function (req, res) {
+  res.sendFile(path.join(__dirname, 'index.html'));
+});
+
+app.listen(port, () => {
+  console.log(`Server is running at http://localhost:${port}`);
+});
```
Comment (chunloy) [inline express.js:29]: This doesn't add value in my opinion

File: index.html [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Document</title>
+    <script src="https://cdn.jsdelivr.net/npm/openseadragon@4.1/build/openseadragon/openseadragon.min.js"></script>
+    <link rel="stylesheet" type="text/css" href="style.css" />
+  </head>
+  <body>
+    <div id="osd" style="width: 100%; height: 405px">
```
Comment (chunloy) [inline index.html:11]: You have styling for this div in your stylesheet, do you need the inline styling?

File: index.html [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+      <div id="presets"></div>
+      <p id="zoomLvl"></p>
```
Comment (chunloy) [inline index.html:13]: Nit: these are hardcoded, could maybe add them programatically.

File: index.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    'https://learn.mghpathology.org/pv-http/WSI22-98-Case-5_1.mrxs.dzi',
+
+  minZoomPixelRatio: 2,
+});
+
+const zoom = document.getElementById('zoomLvl');
+const presets = document.getElementById('presets');
+
+const displayZoom = imagezoom => {
+  return 'Zoom Level: ' + (imagezoom * 20).toFixed(2) + 'x';
```
Comment (chunloy) [inline index.js:21]: Magic number -> factor out to constant.
Not a fan of type coercion, e.g., you're adding strings with numbers. 

Nit: Would be easier to read using string interpolation

File: index.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+};
+
+const displayImage = zoomLvl => {
+  return zoomLvl.slice(0, zoomLvl.length - 1) / 20;
```
Comment (chunloy) [inline index.js:25]: Nit: could use regex, parseFloat. Also, type coercion going on here.

File: index.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+};
+
+const attachedEventListeners = () => {
+  for (let i = 0; i < zoom_Levels.length; i++) {
+    const [img, display] = zoom_Levels[i];
+    const newButton = document.createElement('button');
+    newButton.textContent = `${display}`;
```
Comment (chunloy) [inline index.js:32]: the display variable is already a string, no need to interpolate.

**Review by chunloy — COMMENTED** [review]
This is a good start, nice work. I'd recommend adding some comments.

File: index.js (hunk 6 above)
Comment (jekiquin) [reply index.js:25]: Nit: better function name. What you're trying to do here is convert the display value to zoom level. `displayImage` doesn't seem to say that

File: index.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+const displayZoom = imagezoom => {
+  return 'Zoom Level: ' + (imagezoom * 20).toFixed(2) + 'x';
+};
+
+const displayImage = zoomLvl => {
+  return zoomLvl.slice(0, zoomLvl.length - 1) / 20;
+};
+
+const attachedEventListeners = () => {
```
Comment (jekiquin) [inline index.js:28]: same here. Better function name.  Inside this function, you're currently creating buttons and attaching event listeners

File: index.js (hunk 7 above)
Comment (jekiquin) [inline index.js:32]: it would be better to add a className for the button elements to add more specificity - both on css and js

**Review by jekiquin — COMMENTED** [review]
I'm not sure how detailed the reviewer will be checking.
1. I would do the file structure better
2. Anything that can be factored out as a constant, probably should be factored out as a constant

Before submission
1. Add a demo
2. Be sure you can justify what you have currently done here (why you designed it this way and chose this approach)
3. Submit a clean repo, without this PR. You could also submit a zip file, if ever. I'm not sure what submission they are expecting

File: index.html (hunk 3 above)
Comment (kson1128) [reply index.html:11]: deleted inline and fixed

File: index.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
+const zoom_Levels = [
+  ['2', '40x'],
+  ['1', '20x'],
+  ['0.5', '10x'],
+  ['0.2', '4x'],
+  ['0.05', '1x'],
+];
```
[the remark this answers is not in the document]
Comment (kson1128) [reply index.js:7]: refactored to flattened array

File: index.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  return 'Zoom Level: ' + (imagezoom * 20).toFixed(2) + 'x';
+};
+
+const displayImage = zoomLvl => {
+  return zoomLvl.slice(0, zoomLvl.length - 1) / 20;
+};
+
+const attachedEventListeners = () => {
+  for (let i = 0; i < zoom_Levels.length; i++) {
+    const [img, display] = zoom_Levels[i];
```
[the remark this answers is not in the document]
Comment (kson1128) [reply index.js:30]: renamed function, removed unused variable

File: index.js (hunk 7 above)
Comment (kson1128) [reply index.js:32]: fixed

File: index.js (hunk 8 above)
Comment (kson1128) [reply index.js:28]: fixed

File: index.js (hunk 5 above)
Comment (kson1128) [reply index.js:21]: fixed

File: index.js (hunk 6 above)
Comment (kson1128) [reply index.js:25]: used parseInt and changed name of function

**Pull request closed without merging**